finality-aleph/src/network/manager/compatibility.rs

```diff
@@ -41,21 +41,25 @@ impl<M: Multiaddress> From<DiscoveryMessage<M>> for VersionedAuthentication<M> {
     }
 }
 
-fn encode_with_version(version: Version, mut payload: Vec<u8>) -> Vec<u8> {
-    let mut result =
-        Vec::with_capacity(size_of::<Version>() + size_of::<ByteCount>() + payload.len());
-    version.encode_to(&mut result);
+fn encode_with_version(version: Version, payload: &[u8]) -> Vec<u8> {
     // If size is bigger then u16 we set it to MAX_AUTHENTICATION_SIZE.
     // This should never happen but in case it does we will not panic.
     // Also for other users if they have this version of protocol, authentication
     // will be decoded. If they do not know the protocol, authentication will result
     // in decoding error.
-    payload
+    // We do not have a guarantee that size_hint is implemented for DiscoveryMessage, so we need
+    // to compute actual size to place it in the encoded data.
```

Comment:
But in the body of `fn size_hint` for `VersionedAuthentication<M>` you do have `V1(data) => data.size_hint()` (where `data` is `DiscoveryMessage`). How are these two different?

Reply:
Well, unfortunatelly this is said in parity_scale_codec docs

> size_hint(&self) -> usize: Gets the capacity (in bytes) required for the encoded data. This is to avoid double-allocation of memory needed for the encoding. It can be an estimate and does not need to be an exact number. If the size is not known, even no good maximum, then we can skip this function from the trait implementation. This is required to be a cheap operation, so should not involve iterations etc.

So we cannot use it here as this can be an estimate. I checked what happens in this example, for DiscoveryMessage and it returns `0`. Here we want exact size, so we can't use size_hint.

Reply:
Looking at https://github.com/paritytech/parity-scale-codec/blob/master/src/codec.rs#L246 -- we probably don't even need to implement `size_hint` if we're overwriting the `fn encode` -- `size_hint` is used only in the default implementation of `Encode::encode` and since we're overwriting it, `size_hint` will never be used.

Reply:
True, we also thought about creating a task to rewrite our code to use `size_hint` together with `encode_to` as it is faster. Still it depends how much work would be needed to have this optimization.

Reply:
I would start first with running some benchmarks and looking at the flamegraphs to see how much time is spent doing the decoding/encoding. I can tell you from experience that it can be A LOT.

Reply:
https://cardinal-cryptography.atlassian.net/browse/A0-1481